<h1 align=center><font size = 3>Segmenting and Clustering Neighborhoods in New York City</font></h1>

### Introduction

WE will convert addresses into their equivalent latitude and longitude values. Also, will use the Foursquare API to explore neighborhoods in New York City. Will use the explore function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters using the k-means clustering algorithm to complete this task. Finally, will use the Folium library to visualize the neighborhoods in New York City and their emerging clusters.

Download all the dependencies that we will need.


In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 1. Download and Explore Dataset


In [2]:
!wget -q -O 'newyork_data.json' https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs/newyork_data.json
print('Data downloaded!')

Data downloaded!


#### Load and explore the data

In [3]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [4]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

Notice how all the relevant data is in the _features_ key, which is basically a list of the neighborhoods. So, let's define a new variable that includes this data.


In [5]:
neighborhoods_data = newyork_data['features']
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

#### Transform the data into a _pandas_ dataframe


The next task is essentially transforming this data of nested Python dictionaries into a _pandas_ dataframe. So let's start by creating an empty dataframe.


In [6]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

In [7]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.

In [13]:
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe#scraping the above table,
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [14]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.


In [15]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


#### Use geopy library to get the latitude and longitude values of New York City.


In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.


In [16]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


#### Create a map of New York with neighborhoods superimposed on top.


In [17]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

**Folium** is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.


However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Manhattan. So let's slice the original dataframe and create a new dataframe of the Manhattan data.


In [18]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

Let's get the geographical coordinates of Manhattan.


In [19]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7896239, -73.9598939.


As we did with all of New York City, let's visualizat Manhattan the neighborhoods in it.


In [20]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.


In [21]:
CLIENT_ID = 'E4STEBIGGMI2GLDKBIGNDD4TBPKXA43PTLZJKO0UF2NFWIJL' # your Foursquare ID
CLIENT_SECRET = 'PFHIGTB3WUUQ1SORCBYL3QQ2NK2QWTIOQ30RF0V4PDWDQ1IZ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: E4STEBIGGMI2GLDKBIGNDD4TBPKXA43PTLZJKO0UF2NFWIJL
CLIENT_SECRET:PFHIGTB3WUUQ1SORCBYL3QQ2NK2QWTIOQ30RF0V4PDWDQ1IZ


#### Let's explore the first neighborhood in our dataframe.


In [22]:
manhattan_data.loc[0, 'Neighborhood']

'Marble Hill'

Get the neighborhood's latitude and longitude values.


In [23]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Marble Hill are 40.87655077879964, -73.91065965862981.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.


First, let's create the GET request URL. Name your URL **url**.


In [24]:
radius = 500
limit = 100
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, neighborhood_latitude, neighborhood_longitude, VERSION, radius, LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?client_id=E4STEBIGGMI2GLDKBIGNDD4TBPKXA43PTLZJKO0UF2NFWIJL&client_secret=PFHIGTB3WUUQ1SORCBYL3QQ2NK2QWTIOQ30RF0V4PDWDQ1IZ&ll=40.87655077879964,-73.91065965862981&v=20180605&radius=500&limit=100'

Send the GET request and examine the resutls


In [25]:

# send GET request and get trending venues
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '60f9b88dd490824a4c199ade'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Marble Hill',
  'headerFullLocation': 'Marble Hill, New York',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 23,
  'suggestedBounds': {'ne': {'lat': 40.88105078329964,
    'lng': -73.90471933917806},
   'sw': {'lat': 40.87205077429964, 'lng': -73.91659997808156}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b4429abf964a52037f225e3',
       'name': "Arturo's",
       'location': {'address': '5198 Broadway',
        'crossStreet': 'at 225th St.',
        'lat': 40.87441177110231,
        'lng': -73.91027100981574,
        'labeledLatLngs': [{'label'

we know that all the information is in the items key. Before we proceed, let's borrow the get_category_type function

In [26]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a _pandas_ dataframe.


In [27]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-27-561c05f0fdd1>:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name   categories        lat        lng
0       Arturo's  Pizza Place  40.874412 -73.910271
1    Bikram Yoga  Yoga Studio  40.876844 -73.906204
2  Tibbett Diner        Diner  40.880404 -73.908937
3        Dunkin'   Donut Shop  40.877136 -73.906666
4      Starbucks  Coffee Shop  40.877531 -73.905582

And how many venues were returned by Foursquare?


In [28]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

23 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Manhattan


First, let's create the GET request URL. Name your URL **url**.


#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan


In [29]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [30]:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )

Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


#### Let's check the size of the resulting dataframe


In [27]:
print(manhattan_venues.shape)
manhattan_venues.head()

(3252, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066        Dunkin'   
3  Marble Hill              40.876551               -73.91066  Tibbett Diner   
4  Marble Hill              40.876551               -73.91066      Starbucks   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.877136       -73.906666     Donut Shop  
3       40.880404       -73.908937          Diner  
4       40.877531       -73.905582    Coffee Shop

Let's check how many venues were returned for each neighborhood


In [28]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Battery Park City                       90                      90     90   
Carnegie Hill                           89                      89     89   
Central Harlem                          45                      45     45   
Chelsea                                100                     100    100   
Chinatown                              100                     100    100   
Civic Center                           100                     100    100   
Clinton                                100                     100    100   
East Harlem                             38                      38     38   
East Village                           100                     100    100   
Financial District                     100                     100    100   
Flatiron                               100                     100    100   
Gramercy                                95                      95     95   
Greenwich Village                      100                     100    100   
Hamilton Heights                        61                      61     61   
Hudson Yards                            77                      77     77   
Inwood                                  55                      55     55   
Lenox Hill                             100                     100    100   
Lincoln Square                          98                      98     98   
Little Italy                           100                     100    100   
Lower East Side                         48                      48     48   
Manhattan Valley                        46                      46     46   
Manhattanville                          45                      45     45   
Marble Hill                             23                      23     23   
Midtown                                100                     100    100   
Midtown South                          100                     100    100   
Morningside Heights                     42                      42     42   
Murray Hill                            100                     100    100   
Noho                                   100                     100    100   
Roosevelt Island                        28                      28     28   
Soho                                   100                     100    100   
Stuyvesant Town                         21                      21     21   
Sutton Place                           100                     100    100   
Tribeca                                 92                      92     92   
Tudor City                              81                      81     81   
Turtle Bay                             100                     100    100   
Upper East Side                         98                      98     98   
Upper West Side                         95                      95     95   
Washington Heights                      85                      85     85   
West Village                           100                     100    100   
Yorkville                              100                     100    100   

                     Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                          
Battery Park City                90               90              90  
Carnegie Hill                    89               89              89  
Central Harlem                   45               45              45  
Chelsea                         100              100             100  
Chinatown                       100              100             100  
Civic Center                    100              100             100  
Clinton                         100              100             100  
East Harlem                      38               38              38  
East Village                    100              100             100  
Finan

#### Let's find out how many unique categories can be curated from all the returned venues


In [29]:
print('There are {} uniques categories.'.aformat(len(manhattan_venues['Venue Category'].unique())))

There are 334 uniques categories.


## 3. Analyze Each Neighborhood


In [30]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arepa Restaurant  \
0                   0                    0             0                 0   
1                   0                    0             0                 0   
2                   0                    0             0                 0   
3                   0                    0             0                 0   
4                   0                    0             0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auditorium  Australian Restaurant  \
0                 0                   0           0                      0   
1                 0                   0           0                      0   
2                 0                   0           0                      0   
3                 0                   0           0                      0   
4                 0                   0           0                      0   

   Austrian Restaurant  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                    0          0           0           0       0     0    0   
1                    0          0           0           0       0     0    0   
2                    0          0           0           0       0     0    0   
3                    0          0           0           0       0     0    0   
4                    0          0           0           0       0     0    0   

   Baseball Field  Basketball Court  Beach Bar  Beer Bar  Beer Garden  \
0               0                 0          0         0            0   
1               0                 0          0         0            0   
2               0                 0          0         0            0   
3               0                 0          0         0            0   
4               0                 0          0         0            0   

   Beer Store  Big Box Store  Bike Rental / Bike Share  Bike Trail  Bistro  \
0           0              0                         0           0       0   
1           0              0                         0           0       0   
2           0              0                         0           0       0   
3           0              0                         0           0       0   
4           0              0                         0           0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Boxing Gym  \
0           0              0          0         0           0   
1           0              0          0         0           0   
2           0              0          0         0           0   
3           0              0          0         0           0   
4           0              0          0         0           0   

   Brazilian Restaurant  Breakfast Spot  Bridal Shop  Bridge  Bubble Tea Shop  \
0                     0               0            0       0                0   
1                     0               0            0       0                0   
2                     0               0            0       0                0   
3                     0               0            0       0                

And let's examine the new dataframe size.


In [31]:
manhattan_onehot.shape

(3252, 335)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [32]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000        0.000000               0.00   
1         Carnegie Hill           0.000000        0.000000               0.00   
2        Central Harlem           0.000000        0.000000               0.00   
3               Chelsea           0.000000        0.000000               0.00   
4             Chinatown           0.000000        0.000000               0.00   
5          Civic Center           0.000000        0.000000               0.00   
6               Clinton           0.000000        0.000000               0.00   
7           East Harlem           0.000000        0.000000               0.00   
8          East Village           0.000000        0.000000               0.00   
9    Financial District           0.000000        0.000000               0.00   
10             Flatiron           0.000000        0.000000               0.00   
11             Gramercy           0.000000        0.000000               0.00   
12    Greenwich Village           0.000000        0.000000               0.00   
13     Hamilton Heights           0.000000        0.016393               0.00   
14         Hudson Yards           0.000000        0.000000               0.00   
15               Inwood           0.000000        0.000000               0.00   
16           Lenox Hill           0.000000        0.000000               0.01   
17       Lincoln Square           0.000000        0.000000               0.00   
18         Little Italy           0.000000        0.000000               0.00   
19      Lower East Side           0.000000        0.000000               0.00   
20     Manhattan Valley           0.000000        0.000000               0.00   
21       Manhattanville           0.000000        0.000000               0.00   
22          Marble Hill           0.000000        0.000000               0.00   
23              Midtown           0.000000        0.000000               0.00   
24        Midtown South           0.000000        0.000000               0.00   
25  Morningside Heights           0.000000        0.000000               0.00   
26          Murray Hill           0.000000        0.000000               0.00   
27                 Noho           0.000000        0.000000               0.00   
28     Roosevelt Island           0.000000        0.000000               0.00   
29                 Soho           0.000000        0.000000               0.00   
30      Stuyvesant Town           0.000000        0.000000               0.00   
31         Sutton Place           0.000000        0.010000               0.00   
32              Tribeca           0.000000        0.000000               0.00   
33           Tudor City           0.000000        0.000000               0.00   
34           Turtle Bay           0.000000        0.000000               0.00   
35      Upper East Side           0.000000        0.000000               0.00   
36      Upper West Side           0.010526        0.000000               0.00   
37   Washington Heights           0.011765        0.000000               0.00   
38         West Village           0.010000        0.000000               0.00   
39            Yorkville           0.000000        0.000000               0.00   

    African Restaurant  American Restaurant  Antique Shop  Arepa Restaurant  \
0             0.000000             0.011111          0.00          0.000000   
1             0.000000             0.011236          0.00          0.000000   
2             0.066667             0.044444          0.00          0.000000   
3             0.000000             0.030000          0.00          0.000000   
4             0.000000             0.040000          0.00          0.000000   
5             0.000000             0.040000          0.01          0.000000   
6             0.000000             0.040000          0.00          0.000000   
7             0.000000             0.000000          0.00

#### Let's confirm the new size


In [33]:
manhattan_grouped.shape

(40, 335)

#### Let's print each neighborhood along with the top 5 most common venues


In [34]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
            venue  freq
0            Park  0.10
1     Coffee Shop  0.07
2  Clothing Store  0.06
3           Hotel  0.04
4             Gym  0.03


----Carnegie Hill----
         venue  freq
0  Coffee Shop  0.08
1         Café  0.07
2    Wine Shop  0.04
3  Yoga Studio  0.03
4  Pizza Place  0.03


----Central Harlem----
                 venue  freq
0   African Restaurant  0.07
1    French Restaurant  0.04
2  American Restaurant  0.04
3           Public Art  0.04
4   Chinese Restaurant  0.04


----Chelsea----
                venue  freq
0         Coffee Shop  0.06
1         Art Gallery  0.05
2              Bakery  0.05
3      Ice Cream Shop  0.03
4  Seafood Restaurant  0.03


----Chinatown----
                 venue  freq
0   Chinese Restaurant  0.09
1               Bakery  0.08
2         Cocktail Bar  0.05
3  American Restaurant  0.04
4         Dessert Shop  0.04


----Civic Center----
                  venue  freq
0           Coffee Shop  0.07
1  Gym / Fitness C

#### Let's put that into a _pandas_ dataframe


First, let's write a function to sort the venues in descending order.


In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [36]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park           Coffee Shop   
1      Carnegie Hill           Coffee Shop                  Café   
2     Central Harlem    African Restaurant     French Restaurant   
3            Chelsea           Coffee Shop           Art Gallery   
4          Chinatown    Chinese Restaurant                Bakery   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Clothing Store                 Hotel                   Gym   
1             Wine Shop           Yoga Studio           Pizza Place   
2   American Restaurant            Public Art    Chinese Restaurant   
3                Bakery        Ice Cream Shop    Seafood Restaurant   
4          Cocktail Bar   American Restaurant          Dessert Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Women's Store         Boat or Ferry         Memorial Site   
1             Bookstore     French Restaurant  Gym / Fitness Center   
2  Gym / Fitness Center    Seafood Restaurant                   Bar   
3   American Restaurant             Wine Shop                 Hotel   
4          Optical Shop    Salon / Barbershop              Boutique   

  9th Most Common Venue 10th Most Common Venue  
0            Playground            Pizza Place  
1                   Bar         Cosmetics Shop  
2      Tapas Restaurant    Fried Chicken Joint  
3     French Restaurant       Tapas Restaurant  
4           Coffee Shop                    Spa

## 4. Cluster Neighborhoods


Run _k_-means to cluster the neighborhood into 5 clusters.


In [37]:
# set number of clusters
kclusters = 5

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 0, 0, 0, 0, 1, 1, 3, 0, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.


In [38]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               4   
1  Manhattan           Chinatown  40.715618 -73.994279               0   
2  Manhattan  Washington Heights  40.851903 -73.936900               3   
3  Manhattan              Inwood  40.867684 -73.921210               3   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               3   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                   Gym           Coffee Shop        Sandwich Place   
1    Chinese Restaurant                Bakery          Cocktail Bar   
2                  Café                Bakery                  Bank   
3    Mexican Restaurant            Restaurant                  Café   
4           Pizza Place                  Café         Deli / Bodega   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0      Video Game Store        Tennis Stadium              Pharmacy   
1   American Restaurant          Dessert Shop          Optical Shop   
2     Mobile Phone Shop            Restaurant         Grocery Store   
3                Bakery                Lounge  Caribbean Restaurant   
4           Coffee Shop    Mexican Restaurant           Yoga Studio   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                  Bank       Supplement Shop        Ice Cream Shop   
1    Salon / Barbershop              Boutique           Coffee Shop   
2    Chinese Restaurant    Spanish Restaurant           Coffee Shop   
3              Wine Bar    Chinese Restaurant                  Park   
4                Bakery  Caribbean Restaurant          Cocktail Bar   

  10th Most Common Venue  
0            Pizza Place  
1                    Spa  
2       Tapas Restaurant  
3            Pizza Place  
4                   Park

Finally, let's visualize the resulting clusters


In [39]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters


Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.


#### Cluster 1


In [40]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
1           Chinatown    Chinese Restaurant                   Bakery   
5      Manhattanville           Coffee Shop            Deli / Bodega   
6      Central Harlem    African Restaurant        French Restaurant   
9           Yorkville    Italian Restaurant                      Bar   
10         Lenox Hill    Italian Restaurant              Coffee Shop   
12    Upper West Side    Italian Restaurant                   Bakery   
17            Chelsea           Coffee Shop              Art Gallery   
18  Greenwich Village    Italian Restaurant           Clothing Store   
19       East Village                   Bar              Pizza Place   
20    Lower East Side           Art Gallery       Chinese Restaurant   
21            Tribeca   American Restaurant                     Park   
22       Little Italy                Bakery                     Café   
23               Soho        Clothing Store                 Boutique   
24       West Village    Italian Restaurant  New American Restaurant   
25   Manhattan Valley           Coffee Shop                      Bar   
27           Gramercy    Italian Restaurant                      Bar   
30      Carnegie Hill           Coffee Shop                     Café   
31               Noho          Cocktail Bar       Italian Restaurant   
34       Sutton Place    Italian Restaurant     Gym / Fitness Center   
35         Turtle Bay    Italian Restaurant              Coffee Shop   

   3rd Most Common Venue   4th Most Common Venue     5th Most Common Venue  \
1           Cocktail Bar     American Restaurant              Dessert Shop   
5     Italian Restaurant                     Bar        Seafood Restaurant   
6    American Restaurant              Public Art        Chinese Restaurant   
9            Coffee Shop                     Gym             Deli / Bodega   
10      Sushi Restaurant            Cocktail Bar                      Café   
12              Wine Bar             Coffee Shop                       Bar   
17                Bakery          Ice Cream Shop        Seafood Restaurant   
18              Boutique        Sushi Restaurant            Cosmetics Shop   
19             Speakeasy      Mexican Restaurant            Ice Cream Shop   
20                  Café             Pizza Place          Ramen Restaurant   
21    Italian Restaurant                    Café                  Wine Bar   
22    Italian Restaurant      Chinese Restaurant  Mediterranean Restaurant   
23    Italian Restaurant              Shoe Store  Mediterranean Restaurant   
24                  Park     American Restaurant              Cocktail Bar   
25    Mexican Restaurant   Vietnamese Restaurant           Thai Restaurant   
27           Pizza Place              Bagel Shop                 Wine Shop   
30             Wine Shop             Yoga Studio               Pizza Place   
31                 Hotel      Mexican Restaurant               Pizza Place   
34           Pizza Place  Furniture / Home Store                      Park   
35      Sushi Restaurant        Ramen Restaurant             Deli / Bodega   

        6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1                Optical Shop    Salon / Barbershop              Boutique   
5          Mexican Restaurant    Chinese Restaurant           Supermarket   
6        Gym / Fitness Center    Seafood Restaurant                   Bar   
9            Sushi Restaurant        Ice Cream Shop             Wine Shop   
10                Pizza Place          Burger Joint                   Gym   
12                       Café            Bagel Shop     French Restaurant   
17        American Restaurant             Wine Shop                 Hotel   
18                Coffee Shop          Dessert Shop     Indian Restaurant   
19         Italian Restaurant          Cocktail Bar     Korean Restaurant   
20  Latin American Restaurant                Bakery  Caribbean Restaurant   
21        

#### Cluster 2


In [41]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
8      Upper East Side               Exhibit       Italian Restaurant   
14             Clinton    Italian Restaurant                  Theater   
15             Midtown                 Hotel              Coffee Shop   
16         Murray Hill                 Hotel              Coffee Shop   
29  Financial District           Coffee Shop              Pizza Place   
32        Civic Center           Coffee Shop     Gym / Fitness Center   
33       Midtown South     Korean Restaurant                    Hotel   
38            Flatiron    Italian Restaurant  New American Restaurant   
39        Hudson Yards  Gym / Fitness Center                    Hotel   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
8    American Restaurant  Gym / Fitness Center                Bakery   
14  Gym / Fitness Center        Sandwich Place   American Restaurant   
15        Clothing Store   Sporting Goods Shop             Bookstore   
16   American Restaurant        Sandwich Place   Japanese Restaurant   
29          Cocktail Bar                   Bar                  Park   
32                   Spa     French Restaurant   American Restaurant   
33          Dessert Shop   Japanese Restaurant        Cosmetics Shop   
38   American Restaurant   Japanese Restaurant           Yoga Studio   
39   American Restaurant    Italian Restaurant           Coffee Shop   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
8                  Juice Bar           Coffee Shop                   Spa   
14               Coffee Shop                   Spa                   Gym   
15                   Theater            Steakhouse        Sandwich Place   
16                Taco Place          Burger Joint      Sushi Restaurant   
29                Steakhouse   American Restaurant           Event Space   
32               Yoga Studio    Italian Restaurant                  Park   
33               Coffee Shop             Hotel Bar  Gym / Fitness Center   
38  Mediterranean Restaurant   Sporting Goods Shop             Wine Shop   
39                      Café       Thai Restaurant            Restaurant   

   9th Most Common Venue 10th Most Common Venue  
8         Cosmetics Shop                  Hotel  
14          Cocktail Bar            Pizza Place  
15     Indian Restaurant     Italian Restaurant  
16  Gym / Fitness Center                    Bar  
29    Italian Restaurant                  Hotel  
32          Cocktail Bar                  Hotel  
33   American Restaurant            Karaoke Bar  
38    Mexican Restaurant                    Spa  
39                  Park              Nightclub

#### Cluster 3


In [42]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town                   Bar                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37           Coffee Shop         Boat or Ferry          Cocktail Bar   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37        Farmers Market     German Restaurant           Gas Station   

   9th Most Common Venue 10th Most Common Venue  
37        Baseball Field   Gym / Fitness Center

#### Cluster 4


In [43]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
2    Washington Heights                  Café                Bakery   
3                Inwood    Mexican Restaurant            Restaurant   
4      Hamilton Heights           Pizza Place                  Café   
7           East Harlem    Mexican Restaurant       Thai Restaurant   
11     Roosevelt Island           Coffee Shop                  Park   
13       Lincoln Square                 Plaza                  Café   
26  Morningside Heights             Bookstore           Coffee Shop   
28    Battery Park City                  Park           Coffee Shop   
36           Tudor City    Mexican Restaurant                  Park   

   3rd Most Common Venue      4th Most Common Venue  5th Most Common Venue  \
2                   Bank          Mobile Phone Shop             Restaurant   
3                   Café                     Bakery                 Lounge   
4          Deli / Bodega                Coffee Shop     Mexican Restaurant   
7                 Bakery  Latin American Restaurant          Deli / Bodega   
11     Kosher Restaurant            Bubble Tea Shop            Supermarket   
13          Concert Hall                    Theater  Performing Arts Venue   
26                  Park        American Restaurant          Deli / Bodega   
28        Clothing Store                      Hotel                    Gym   
36                  Café                Pizza Place                    Gym   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2          Grocery Store    Chinese Restaurant    Spanish Restaurant   
3   Caribbean Restaurant              Wine Bar    Chinese Restaurant   
4            Yoga Studio                Bakery  Caribbean Restaurant   
7         Sandwich Place                   Spa         Grocery Store   
11                   Gym              Bus Line               Dog Run   
13             Wine Shop            Food Truck           Coffee Shop   
26        Sandwich Place          Burger Joint                  Café   
28         Women's Store         Boat or Ferry         Memorial Site   
36                 Diner           Coffee Shop      Greek Restaurant   

    9th Most Common Venue 10th Most Common Venue  
2             Coffee Shop       Tapas Restaurant  
3                    Park            Pizza Place  
4            Cocktail Bar                   Park  
7      Seafood Restaurant             Taco Place  
11  Outdoors & Recreation           Soccer Field  
13                   Park   Gym / Fitness Center  
26            Pizza Place         Ice Cream Shop  
28             Playground            Pizza Place  
36     Seafood Restaurant                 Garden

#### Cluster 5


In [44]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Marble Hill                   Gym           Coffee Shop   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Sandwich Place      Video Game Store        Tennis Stadium   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0              Pharmacy                  Bank       Supplement Shop   

  9th Most Common Venue 10th Most Common Venue  
0        Ice Cream Shop            Pizza Place

End

#### Cluster 5